First pass at editor GUI

kanerogers opened the pull request:
## What is this related to?
Closes #419 

## What's changed?
- Began work on new, amazing, perfect and flawless OpenXR runtime (TM)
- Created stub app for editor
- Created a "editor" feature flag to keep @rasmusgo hap^H^H^H^H^H^H^Hencourage good development practices

kanerogers committed:
Begin work on editor

- Create stub app for Editor
- Begin work on new OpenXR client
- Create "editor" feature flag to hide new work

rasmusgo commented on hotham-editor/src/main.rs:
```diff
@@ -0,0 +1,362 @@
+                .unwrap();
+
+            let handle = external_semaphore
+                .get_semaphore_win32_handle(
+                    &vk::SemaphoreGetWin32HandleInfoKHR::builder()
```
😨 only windows?

kanerogers replied:
Nope! This can be done on each platform, but each platform has its own unique way of doing it. Focussing on Windows for the first pass.

rasmusgo commented on hotham/src/contexts/xr_context/mod.rs:
```diff
@@ -464,6 +464,10 @@ fn enable_xr_extensions(required_extensions: &mut xr::ExtensionSet) {
 fn enable_xr_extensions(required_extensions: &mut xr::ExtensionSet) {
     required_extensions.khr_vulkan_enable = true;
+
+    if cfg!(feature = "editor") {
```
I there any reason to use `if cfg!(..` instead of `#[cfg(..`?

kanerogers replied:
Can't do attributes on statements, apparently. 🤷

rasmusgo commented on hotham-openxr-client/src/lib.rs:
```diff
@@ -0,0 +1,163 @@
+
+    let name = CStr::from_ptr(name);
+    let name = name.to_bytes();
+    if name == b"xrGetInstanceProcAddr" {
```
Could this be a match expression instead?

kanerogers replied:
Eh. I've tried a couple different implementations of this - the if statements are messy and yuck, but they work.

kanerogers committed:
Add `hotham-editor-protocol`

- Define a way for the editor and clients to talk to eachother
- Simple request/response protocol that's transport agnostic
- Works perfectly and has no flaws

kanerogers committed:
Actually, no bytemuck yet due to dependency hell

kanerogers committed:
Get images transferring from server to client

kanerogers committed:
Finish first pass at editor input

kanerogers committed:
Improve input handling

kanerogers committed:
Fix up input

kanerogers committed:
Begin adding support for JSON values

kanerogers committed:
WIP

rasmusgo commented on .vscode/settings.json:
```diff
@@ -6,5 +6,9 @@
     "files.trimTrailingWhitespace": true,
-    "webgl-glsl-editor.format.placeSpaceAfterKeywords": true
+    "webgl-glsl-editor.format.placeSpaceAfterKeywords": true,
+    "rust-analyzer.check.extraArgs": [
+        "--target-dir",
+        "C:/windows/temp/rust-analyzer-check"
+    ],
```
This doesn't look portable

kanerogers replied:
Yeah, I'm thinking I should just remove this file from the repo

rasmusgo replied:
I'm not convinced the best way forward is to remove the whole file. These settings are good for keeping the files tidy and consistent. Can we at least resolve this in a separate PR?

kanerogers replied:
I guess the problem is that there's always going to be a case where a developer might have overrides that are unique to their specific environment, right? For me the rust analyzer one is super common that I need *per project* to avoid locking issues.

I think anything related to style etc should probably best be left to clippy, otherwise there's no way to, for example. get developers who aren't using VSCode to have their code correctly formatted.

rasmusgo commented on hotham-editor/src/main.rs:
```diff
@@ -0,0 +1,418 @@
+use crate::camera::Camera;
+
+/// Compile your own damn shaders! LazyVulkan is just as lazy as you are!
+static FRAGMENT_SHADER: &'_ [u8] = include_bytes!("shaders/triangle.frag.spv");
+static VERTEX_SHADER: &'_ [u8] = include_bytes!("shaders/triangle.vert.spv");
```
These files are not checked in

kanerogers replied:
Ah! Good pickup.

rasmusgo replied:
You may want to add a `.gitignore` with exceptions for these files. Checking in files that are ignored by git can lead to confusing stuff, eg. local changes are not picked up by git even though the file is tracked.

kanerogers replied:
Hrm, I've found that it's usually fine after you've explicitly added that file to the index, but I've added an exception to the gitignore just to be on the safe side

kanerogers committed:
[Editor] Finish first pass at editor GUI
Fixes #419

kanerogers committed:
Add missing SPRIV files

kanerogers committed:
Appease clippy

kanerogers committed:
Merge branch 'main' into kanerogers/issue419

kanerogers committed:
Specifically allow the editor's SPIRV files

kanerogers committed:
Merge branch 'kanerogers/issue419' of https://github.com/leetvr/hotham into kanerogers/issue419

rasmusgo commented on hotham-editor/src/shaders/triangle.vert:
```diff
@@ -0,0 +1,14 @@
+#version 450
+
+layout(location = 0) in vec4 in_position;
+layout(location = 1) in vec4 in_colour;
```
"color"

kanerogers replied:
![scowl](https://media3.giphy.com/media/xgS0aDRPnHTNt0PMKz/giphy.gif?cid=ecf05e47uxz8svws1zr7t6bl2me6vepol22k9s3popihxs45&rid=giphy.gif&ct=g)

kanerogers committed:
Accept the tyrany of The Republic

kanerogers commented:
Need to fix the dependencies so we're not pulling in UDS on Linux.

louisgjohnson commented on hotham-openxr-client/src/lib.rs:
```diff
@@ -0,0 +1,158 @@
+    } else {
+        let _name = String::from_utf8_unchecked(name.to_vec());
+        unsafe extern "system" fn bang() -> Result {
+            panic!("UNIMPLEMENTED FUNCTION!");
```
Could potentially add the name to the panic to make it easier to track down.

kanerogers replied:
This is a fun one! So I've tried to pass this in but for whatever reason - potentially FFI weirdness, I am clearly not clever enough to understand those machinations - it isn't possible to return a function that will panic with the correct function name.

kanerogers committed:
Merge branch 'main' into kanerogers/issue419

kanerogers committed:
WIP: Fixing up state issues

kanerogers committed:
Clean up the OpenXR client a bit
